### PR TITLE
Partial multi selection: limit if selection contains unmergeable block

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Even though the selection could be mergeable (if the start and end are), it may not be what most users may intent to select at that point.

Needs design help: currently only the borders are visible, but we need to add a background colour so the native selection disappears.

### CODE REVIEW

File: packages/block-editor/src/store/selectors.js [hunk 1]
```diff
@@ -949,6 +949,14 @@ export function __unstableIsSelectionCollapsed( state ) {
 	);
 }
 
+export function __unstableSelectionHasUnmergeableBlock( state ) {
+	return getSelectedBlockClientIds( state ).some( ( clientId ) => {
+		const blockName = getBlockName( state, clientId );
+		const blockType = getBlockType( blockName );
+		return ! blockType.merge;
```
Comment (ellatrix) [inline packages/block-editor/src/store/selectors.js:955]: May need to fine tune this by also checking if `switchToBlockType` returns a result.

**Comment by jasmussen** [comment]
Thank you for the PR! Overwhelmingly when I make partial selections between two text blocks it is to delete just the selection and merge them. And when I select text and image, on the other hand, it's to delete both blocks in their entirety. 

As exists in trunk at the moment, the selection doesn't always look great when made across multiple complex blocks, especially in the site editor. I think this PR can thread the needle:

status

I'm going to dive into see what we can do, stylewise, when flipping between partial and multi selection, explore what options we have to make that look good. 

> Needs design help: currently only the borders are visible, but we need to add a background colour so the native selection disappears.

Just as an experiment, does this code work for you?

```
::selection {
	background: transparent;
}
```

It would need to be contextual to just when blocks are multi selected (perhaps the hard part?), but theoretically it would allow us more control over the multi-selection style. Just as an experiment here's what I tried in the inspector:
test

**Comment by ellatrix** [comment]
> Just as an experiment, does this code work for you?

Could you try it in this PR? It doesn't work for me to flip between visible/invisible while the selection is being set.

> As exists in trunk at the moment, the selection doesn't always look great when made across multiple complex blocks, especially in the site editor.

That's because we look at native selection. What we can do it also look at `mouseenter` for blocks perhaps?

**Comment by jasmussen** [comment]
Diving in a bit deeper, this is what I'm seeing now:
currently

I see the behavior: onmouseup, the native selection is hidden. Makes sense.

Regardless of what we can do with the multi selection style (I'll dive into some mockups in a moment), one thing that could be worth looking at, is _when_ the multi selection applies. It is partially visible in the GIF above: I would expect the multi-selection to apply as soon as I drag into the image, but that doesn't always happen. In some cases, the selection only applies after I drag, what feels like, half way through the image. Do you have any idea if this is only on my side or not?

Nice work.

**Comment by ellatrix** [comment]
> It is partially visible in the GIF above: I would expect the multi-selection to apply as soon as I drag into the image, but that doesn't always happen.

That's what I meant before:

> That's because we look at native selection. What we can do it also look at `mouseenter` for blocks perhaps?

It's just how the browser is handling multi selection, but I guess we can look at additional events.

**Comment by jasmussen** [comment]
A couple of sketches of how selection could look. The following I would think to be the idealized version:

singular selection

Essentially the same color that's used for the partial selection is used as a singular overlay color for the whole multi selection. I think this is likely going to be a bit of a technical challenge which I'll try and tinker with in a codepen. Ensuring contrast using filters is an idea, but it's a question how well that will work on different backgrounds. Unfortunately `background-color: Highlight;` has been deprecated, otherwise that would've been ideal.

Another option is to keep a plain partial selection style, and keep a customized multi-select style. In that case, I'm wondering if we shouldn't keep the spot color blue, including the border, but add a faded overlay as well. What do you think?

Selecting, blue bg

**Comment by jasmussen** [comment]
A couple of quick codepens and thoughts to get the discussion going. First off, [this codepen](https://codepen.io/joen/pen/RwMMMPX), with **native select, custom multi select**:

native select, custom multi select

And with **custom select and multi select**:
custom select and multi select

Neither of these are ideal, as the two selection styles feel too far removed from each other.

Another custom version, without the overlay:

custom select and multiselect 2

same, but with native select:

custom select and multiselect 3

These two are close to what we have today, but also not home-runs as the multi select style isn't necessarily as clear without the overlay.

[This codepen](https://codepen.io/joen/pen/eYMMrEw?editors=1100) I _visually_ prefer, as it visually appears as a truly coherent singular style of selecting, both partial and multi. The colors/opacity aren't quite right, but the principle feels solid to me. **Custom partial and multiselect:**

alternate custom select

This last one presents a technical challenge, though, in terms of ensuring it works in all contexts.

**Comment by ellatrix** [comment]
@jasmussen I actually had it wrong. It's not the fault of the native selection. We were looking at it wrong. I fixed it in https://github.com/WordPress/gutenberg/pull/42983.

**Comment by jasmussen** [comment]
I'd still love to land a version of this one!

**Comment by ellatrix** [comment]
Okay, what do we still need for this one? :)

**Review by jasmussen — APPROVED** [review]
This is trunk:

before

Notice how when you select into a placeholder, there's actually text being selected inside which fiddles with the multi selection.

This branch marks a substantially clearer experience:

after

As a move forward, this feels like a great move. It doesn't prevent us from future iteration, but it improves what's shipping. Thank you!

I'd love a code sanity check on this one. But the experience is much stronger.

**Comment by ellatrix** [comment]
@jasmussen You didn't manage to remove the native selection when it goes over into full selection, right?

In any case, let's ship this and iterate.

**Comment by jasmussen** [comment]
> You didn't manage to remove the native selection when it goes over into full selection, right?

It's invisible. I don't think `user-select: none;` would work either — I seem to recall testing that and it still behaving the same — but that's an option we can look at if it continues.

**Comment by adamziel** [comment]
Please take a look at https://github.com/WordPress/gutenberg/pull/44521 – it makes the new experimental/unstable selectors introduced in this PR private.

**Pull request merged**